cmd/builder/internal/builder/main.go

```diff
@@ -131,29 +144,74 @@ func GetModules(cfg Config) error {
 	}
 
 	// ambiguous import: found package cloud.google.com/go/compute/metadata in multiple modules
-	if err := runGoCommand(cfg, "get", "cloud.google.com/go"); err != nil {
-		return fmt.Errorf("failed to go get: %w", err)
+	if _, err := runGoCommand(cfg, "get", "cloud.google.com/go"); err != nil {
+		return fmt.Errorf("%w: %v", errGoGetFailed, err)
 	}
 
-	if err := runGoCommand(cfg, "mod", "tidy", "-compat=1.21"); err != nil {
+	if _, err := runGoCommand(cfg, "mod", "tidy", "-compat=1.21"); err != nil {
 		return fmt.Errorf("failed to update go.mod: %w", err)
 	}
 
+	if cfg.SkipStrictVersioning {
+		return downloadModules(cfg)
+	}
+
+	// Perform strict version checking.  For each component listed and the
+	// otelcol core dependency, check that the enclosing go module matches.
+	modulePath, dependencyVersions, err := cfg.readGoModFile()
+	if err != nil {
+		return err
+	}
+
+	corePath, coreVersion := cfg.coreModuleAndVersion()
+	coreDepVersion, ok := dependencyVersions[corePath]
+	if !ok {
+		return fmt.Errorf("core collector %w: '%s'. %s", ErrDepNotFound, corePath, skipStrictMsg)
+	}
+	if coreDepVersion != coreVersion {
+		return fmt.Errorf(
+			"%w: core collector version calculated by component dependencies %q does not match configured version %q. %s",
+			ErrVersionMismatch, coreDepVersion, coreVersion, skipStrictMsg)
+	}
+
+	for _, mod := range cfg.allComponents() {
+		module, version, _ := strings.Cut(mod.GoMod, " ")
+		if module == modulePath {
+			// No need to check the version of components that are part of the
+			// module we're building from.
+			continue
+		}
+
+		moduleDepVersion, ok := dependencyVersions[module]
+		if !ok {
+			return fmt.Errorf("component %w: '%s'. %s", ErrDepNotFound, module, skipStrictMsg)
+		}
+		if moduleDepVersion != version {
+			return fmt.Errorf(
+				"%w: component %q version calculated by dependencies %q does not match configured version %q. %s",
+				ErrVersionMismatch, module, moduleDepVersion, version, skipStrictMsg)
+		}
+	}
+
+	return downloadModules(cfg)
+}
+
+func downloadModules(cfg Config) error {
 	cfg.Logger.Info("Getting go modules")
 	// basic retry if error from go mod command (in case of transient network error). This could be improved
 	// retry 3 times with 5 second spacing interval
 	retries := 3
 	failReason := "unknown"
 	for i := 1; i <= retries; i++ {
-		if err := runGoCommand(cfg, "mod", "download"); err != nil {
+		if _, err := runGoCommand(cfg, "mod", "download"); err != nil {
```

Comment:
I haven't been able to add coverage for this because if I would expect this to fail, the previous `go get` [command](https://github.com/open-telemetry/opentelemetry-collector/pull/9897/files#diff-117127e4cf9f6587102dc92a7240d1e2c00944ee47ce23e02731f3a41466bdb1R147) fails first.